### ISSUE TITLE
Retention service hanging during scheduled deletion check / shard deletion

### ISSUE THREAD

**Issue by philippreston** [influxdata/influxdb#11133]
In trying to control disk space with retention policies we are hitting an issue where influx does not seem to be removing expired shards.  So at the minute in our running database I can see the following shards (filtered to one database):

```
24 telegraf   default          2019-01-03T00:00:00Z 2019-01-04T00:00:00Z 2019-01-14T00:00:00Z
27 telegraf   default          2019-01-04T00:00:00Z 2019-01-05T00:00:00Z 2019-01-15T00:00:00Z
30 telegraf   default          2019-01-05T00:00:00Z 2019-01-06T00:00:00Z 2019-01-16T00:00:00Z
33 telegraf   default          2019-01-06T00:00:00Z 2019-01-07T00:00:00Z 2019-01-17T00:00:00Z
36 telegraf   default          2019-01-07T00:00:00Z 2019-01-08T00:00:00Z 2019-01-18T00:00:00Z
40 telegraf   default          2019-01-08T00:00:00Z 2019-01-09T00:00:00Z 2019-01-19T00:00:00Z
43 telegraf   default          2019-01-09T00:00:00Z 2019-01-10T00:00:00Z 2019-01-20T00:00:00Z
46 telegraf   default          2019-01-10T00:00:00Z 2019-01-11T00:00:00Z 2019-01-21T00:00:00Z
49 telegraf   default          2019-01-11T00:00:00Z 2019-01-12T00:00:00Z 2019-01-22T00:00:00Z
52 telegraf   default          2019-01-12T00:00:00Z 2019-01-13T00:00:00Z 2019-01-23T00:00:00Z
55 telegraf   default          2019-01-13T00:00:00Z 2019-01-14T00:00:00Z 2019-01-24T00:00:00Z
58 telegraf   default          2019-01-14T00:00:00Z 2019-01-15T00:00:00Z 2019-01-25T00:00:00Z
62 telegraf   default          2019-01-15T00:00:00Z 2019-01-16T00:00:00Z 2019-01-26T00:00:00Z
65 telegraf   default          2019-01-16T00:00:00Z 2019-01-17T00:00:00Z 2019-01-27T00:00:00Z
```

So today the server date is:

```
~ » date                                                                                                                                                        
Wed Jan 16 12:45:50 UTC 2019
```
Given that the expiry_time of shards 24, 27 and 30 have now past - I believe these should have been removed... but have not.  The retention policy for this database is as follows:

```
> use telegraf
Using database telegraf
> show retention policies
name    duration shardGroupDuration replicaN default
----    -------- ------------------ -------- -------
default 240h0m0s 24h0m0s            1        true
```

So as my understanding this will keep 24 hour shards of data, and keep 240 hours worth / 10 days. 

Is there something I am missing here?

Thanks

Phil

**Comment by philippreston** [comment]
Providing a bit more detail on this after digging into the logs.  This actually seems like the retention service is hung on a WaitGroup, regarding disabling level compaction.  

So looking at the logs I can see the following:

```
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.255686Z lvl=info msg="Retention policy deletion check (start)" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check op_event=start
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.264588Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=23 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.286598Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=26 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.289070Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=29 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.290809Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=32 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.292709Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=35 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.294389Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=39 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.296294Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=42 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.297975Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=45 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.299971Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=48 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.301908Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_group=51 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.304082Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_group=36 db_rp=default
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.305737Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_group=40 db_rp=default
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.307706Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_group=43 db_rp=default
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.309398Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_group=46 db_rp=default
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.311355Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_group=49 db_rp=default
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.313031Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_group=52 db_rp=default
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.314955Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_group=34 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.316631Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_group=37 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.318626Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_group=41 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.320335Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_group=44 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.322267Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_group=47 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.324052Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_group=50 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.326141Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_group=53 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.328013Z lvl=info msg="Deleted shard group" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=chronograf db_shard_group=4 db_rp=autogen
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.400548Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_id=34 db_rp=monitor
Jan 20 19:00:36 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:36.414419Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_id=29 db_rp=autogen
Jan 20 19:00:37 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:37.471036Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_id=40 db_rp=default
Jan 20 19:00:37 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:37.776330Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_id=18 db_rp=default
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.283849Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_id=43 db_rp=default
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.327851Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_id=44 db_rp=monitor
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.373393Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_id=53 db_rp=monitor
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.375878Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_id=28 db_rp=monitor
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.384566Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_id=20 db_rp=autogen
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.387212Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_id=47 db_rp=monitor
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.448754Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_id=35 db_rp=autogen
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.471908Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=metamako db_shard_id=48 db_rp=autogen
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.488903Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_id=37 db_rp=monitor
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.491453Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=_internal db_shard_id=50 db_rp=monitor
Jan 20 19:00:38 ch1-tickstack influxd[32306]: ts=2019-01-20T19:00:38.856033Z lvl=info msg="Deleted shard" log_id=0D787D30000 service=retention trace_id=0D7A0q7l000 op_name=retention_delete_check db_instance=telegraf db_shard_id=46 db_rp=default
```

So we are getting the log line Retention policy deletion check (start), but we never see the end.  From looking at the stack trace:

```
1 @ 0x42fcfb 0x42fda3 0x44069c 0x4402d9 0x475ca4 0xeb55d1 0xeb52f2 0xeb848c 0x9a8cd7 0x9a8c49 0x9ba038 0xd1f5f1 0xd21837 0x45d971
#   0x4402d8    sync.runtime_Semacquire+0x38                                /usr/local/go/src/runtime/sema.go:56
#   0x475ca3    sync.(*WaitGroup).Wait+0x63                             /usr/local/go/src/sync/waitgroup.go:130
#   0xeb55d0    github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).disableLevelCompactions+0x110 /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:455
#   0xeb52f1    github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).SetCompactionsEnabled+0x71    /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:383
#   0xeb848b    github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Engine).Close+0x3b            /go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:761
#   0x9a8cd6    github.com/influxdata/influxdb/tsdb.(*Shard).close+0x46                 /go/src/github.com/influxdata/influxdb/tsdb/shard.go:378
#   0x9a8c48    github.com/influxdata/influxdb/tsdb.(*Shard).Close+0x78                 /go/src/github.com/influxdata/influxdb/tsdb/shard.go:368
#   0x9ba037    github.com/influxdata/influxdb/tsdb.(*Store).DeleteShard+0x2c7              /go/src/github.com/influxdata/influxdb/tsdb/store.go:734
#   0xd1f5f0    github.com/influxdata/influxdb/services/retention.(*Service).run+0x260          /go/src/github.com/influxdata/influxdb/services/retention/service.go:134
#   0xd21836    github.com/influxdata/influxdb/services/retention.(*Service).Open.func1+0x56        /go/src/github.com/influxdata/influxdb/services/retention/service.go:51
```

You can see that its waiting on the WaitGroup as part of this function:

```golang
// disableLevelCompactions will stop level compactions before returning.
//
// If 'wait' is set to true, then a corresponding call to enableLevelCompactions(true) will be
// required before level compactions will start back up again.
func (e *Engine) disableLevelCompactions(wait bool) {
   e.mu.Lock()
   old := e.levelWorkers
   if wait {
      e.levelWorkers += 1
   }

   // Hold onto the current done channel so we can wait on it if necessary
   waitCh := e.done
   wg := e.wg

   if old == 0 && e.done != nil {
      // It's possible we have closed the done channel and released the lock and another
      // goroutine has attempted to disable compactions.  We're current in the process of
      // disabling them so check for this and wait until the original completes.
      select {
      case <-e.done:
         e.mu.Unlock()
         return
      default:
      }

      // Prevent new compactions from starting
      e.Compactor.DisableCompactions()

      // Stop all background compaction goroutines
      close(e.done)
      e.mu.Unlock()
      wg.Wait()  <-- WAITING 

      // Signal that all goroutines have exited.
      e.mu.Lock()
      e.done = nil
      e.mu.Unlock()
      return
   }
   e.mu.Unlock()

   // Compaction were already disabled.
   if waitCh == nil {
      return
   }

   // We were not the first caller to disable compactions and they were in the process
   // of being disabled.  Wait for them to complete before returning.
   <-waitCh
   wg.Wait()
}
```

I'll attach the syslog and  profiles.tar.gz which includes the stack trace.  l will continue to look at the code.  but I throw this up in case its something you have seen before.

Thanks

Phil

[profiles.tar.gz](https://github.com/influxdata/influxdb/files/2782591/profiles.tar.gz)
[syslog-influx.log.gz](https://github.com/influxdata/influxdb/files/2782592/syslog-influx.log.gz)

**Comment by philippreston** [comment]
This is resolved when updating to 1.7.3.  Will reopen if reoccurs.
